Add allowed xblocks which an olx may contain but currently not supported

awaisdar001 opened the pull request:
Currently, there are many `xblocks` that a course Olx may contain but they are not supported by `olxcleaner`. This PR adds a robust way to bypass that. 

`olxcleaner.validate` API now allows you to provide a list of supported xblocks that may and may not be part of this library already. e.g. 
```
    allowed_xblocks = [
        "recommender",
        "edx_sga",
        "crowdsourcehinter",
        "done",
        "word_cloud",
        "wiki"
    ]
    course, errorstore, url_names = validate("testcourses/testcourse11", steps=1, allowed_xblocks=allowed_xblocks)
```

Providing a list of xblock as `allowed_xblocks` the library will not throw a `UnexpectedTag` error. 

```
"A <recommender> tag was unexpectedly found inside the <vertical url_name='myvertical6' display_name='Hello'> tag")
"A <wiki> tag was unexpectedly found inside the <course display_name='[Test] RecommenderXBlock'> tag")
"A <chapter> tag was unexpectedly found inside the <vertical> tag"
```

FYI -- @pdpinch @jolyonb

https://openedx.atlassian.net/browse/TNL-8147

jolyonb commented:
@awaisdar001 Makes me so happy to see somebody pick this up! Thank you!

Do you have a reviewer in mind, or would you like my input too?

awaisdar001 commented:
> @awaisdar001 Makes me so happy to see somebody pick this up! Thank you!
> 
> Do you have a reviewer in mind, or would you like my input too?

@jolyonb sure, I am happy if you can provide some input here.

asadazam93 commented on olxcleaner/loader/xml.py:
```diff
@@ -6,37 +6,32 @@
     :param directory: Path for course.xml (or equivalent)
     :param filename: Filename for course.xml (or equivalent)
     :param errorstore: ErrorStore object to store errors
+    :param allowed_xblocks: ErrorStore object to store errors
```
Incorrect parameter description

awaisdar001 replied:
Done.

asadazam93 commented on olxcleaner/loader/xml.py:
```diff
@@ -80,6 +76,7 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
     :param errorstore: An ErrorStore object that is collecting errors
     :param htmlfiles: A dictionary of XML filenames (value) that reference a given HTML filename (key)
     :param pointer: True if we've arrived at this node due to a pointer tag
+    :param allowed_xblocks: True if we've arrived at this node due to a pointer tag
```
Incorrect parameter description

asadazam93 commented on olxcleaner/exceptions.py:
```diff
@@ -59,3 +61,7 @@ def name(self):
+
+
+class ClassDoesNotExist(ValueError):
+    """Error raised when we support an xblock but it's support is not available in library."""
```
Error raised when an xblock is allowed but not supported by the library

asadazam93 commented on olxcleaner/loader/xml.py:
```diff
@@ -239,15 +237,36 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
+    :param allowed_xblocks: List of all allowed xblocks which may or may not be supported currently.
+    """
+    tag_is_allowed = allowed_xblocks is not None and tag in allowed_xblocks
+    if allowed_xblocks is None:
```
If we check this condition first there will be no need to check for empty in the above line.
```
if allowed_xblocks is None:
    return tag in allowed_children
else:
    return tag in allowed_xblocks
```

awaisdar001 replied:
Done.

asadazam93 commented on olxcleaner/loader/xml.py:
```diff
@@ -239,15 +237,36 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
+    Checks if a tag is valid or not.
+
+    :param tag: Tag name of the lxml element
+    :param allowed_children: List of xblocks which this tag can have
```
List of children/xblock allowed by the olxcleaner

asadazam93 commented on olxcleaner/loader/xml.py:
```diff
@@ -239,15 +237,36 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
+
+    :param tag: Tag name of the lxml element
+    :param allowed_children: List of xblocks which this tag can have
+    :param allowed_xblocks: List of all allowed xblocks which may or may not be supported currently.
```
List of Xblocks that are allowed but may not be supported by the olxcleaner for validation

awaisdar001 commented:
@asadazam93 addressed your comments.

jolyonb commented on olxcleaner/loader/xml.py:
```diff
@@ -239,15 +237,35 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
+    """
+    if allowed_xblocks is None:
+        return tag in allowed_children
+    return tag in allowed_xblocks
```
This logic doesn't seem quite right to me. I think you want `return tag in allowed_xblocks or tag in allowed_children`.

awaisdar001 replied:
as `allowed_xblocks` can be None so the above-mentioned code can break.

jolyonb replied:
This is intended to replace line 271, not line 269-270.

awaisdar001 replied:
Ah I see, I understand.

jolyonb commented:
I'm a little concerned that the approach you're taking here allows errors to slip through. For example, if you have `vertical` in `allowed_xblocks`, then your code will allow `vertical` to be a child of any other xblock, including, for example, `vertical`. I suspect this can be fixed by filtering out classes that are already validated from `allowed_xblocks`.

awaisdar001 commented:
> I'm a little concerned that the approach you're taking here allows errors to slip through. For example, if you have `vertical` in `allowed_xblocks`, then your code will allow `vertical` to be a child of any other xblock, including, for example, `vertical`. I suspect this can be fixed by filtering out classes that are already validated from `allowed_xblocks`.

Thanks for the input @jolyonb -- Do you mean to filter out values from `allowed_xblocks` that already exist in `EdxObject.__subclasses__()`?

jolyonb commented:
> Thanks for the input @jolyonb -- Do you mean to filter out values from `allowed_xblocks` that already exist in `EdxObject.__subclasses__()`?

Exactly.

awaisdar001 commented:
> > Thanks for the input @jolyonb -- Do you mean to filter out values from `allowed_xblocks` that already exist in `EdxObject.__subclasses__()`?
> 
> Exactly.

👍 I'll take a look.

awaisdar001 commented:
@jolyonb I have addressed your comments, can you take a look at the following commit?
https://github.com/openedx/olxcleaner/pull/1/commits/fd46d1d4a3e2c8834fe50af7f0b1d80d6b42940f

jolyonb commented:
Looks good to me now! 👍

awaisdar001 commented:
> Looks good to me now! 👍

Thanks @jolyonb 🎆

awaisdar001 commented:
🎆 first Pr in this fork 🎆
